File: java-mvc/hexagonal-architecture/src/main/java/com/ecommerce/hexagonal/domain/service/ProductService.java
package com.ecommerce.hexagonal.domain.service;

import com.ecommerce.hexagonal.domain.model.Product;
import com.ecommerce.hexagonal.domain.port.input.ProductServicePort;
import com.ecommerce.hexagonal.domain.port.output.ProductRepositoryPort;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Domain Service - implements the input port.
 * Contains business logic orchestration.
 * Depends ONLY on domain model and output ports (interfaces).
 * NO framework annotations here - Spring config wires this up.
 */
public class ProductService implements ProductServicePort {

    private final ProductRepositoryPort productRepository;

    public ProductService(ProductRepositoryPort productRepository) {
        this.productRepository = productRepository;
    }

    @Override
    public Product createProduct(String name, String description, BigDecimal price,
                                  String category, Integer stockQuantity, String sku) {
        // Business rule: SKU must be unique
        if (sku != null && productRepository.existsBySku(sku)) {
            throw new IllegalArgumentException("Product with SKU '" + sku + "' already exists");
        }

        Product product = Product.create(name, description, price, category, stockQuantity, sku);
        return productRepository.save(product);
    }

    @Override
    public Optional<Product> getProductById(UUID id) {
        return productRepository.findById(id);
    }

    @Override
    public List<Product> getAllProducts(int page, int size) {
        if (size > 100) {
            size = 100; // Business rule: max page size is 100
        }
        return productRepository.findAll(page, size);
    }

    @Override
    public List<Product> searchProducts(String query, String category,
                                         BigDecimal minPrice, BigDecimal maxPrice) {
        return productRepository.search(query, category, minPrice, maxPrice);
    }

    @Override
    public Product updateProduct(UUID id, String name, String description,
                                  BigDecimal price, String category) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + id));

        product.updateDetails(name, description, price, category);
        return productRepository.save(product);
    }

    @Override
    public void deleteProduct(UUID id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + id));

        product.deactivate();
        productRepository.save(product);
    }

    @Override
    public void decreaseStock(UUID productId, int quantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + productId));

        product.decreaseStock(quantity);
        productRepository.save(product);
    }

    @Override
    public void increaseStock(UUID productId, int quantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + productId));

        product.increaseStock(quantity);
        productRepository.save(product);
    }

    @Override
    public long countProducts() {
        return productRepository.count();
    }
}

File: dotnet-mvc/HexagonalArchitecture/Domain/Services/ProductService.cs
using HexagonalArchitecture.Domain.Models;
using HexagonalArchitecture.Domain.Ports.Input;
using HexagonalArchitecture.Domain.Ports.Output;

namespace HexagonalArchitecture.Domain.Services;

/// <summary>
/// Domain service implementing the input port.
/// This belongs to the DOMAIN layer and contains application/orchestration logic.
/// It has NO framework annotations — it is registered via DI in Program.cs.
/// It depends only on the output port interface (IProductRepository), not on concrete implementations.
/// </summary>
public class ProductService : IProductService
{
    private readonly IProductRepository _repository;

    public ProductService(IProductRepository repository)
    {
        _repository = repository;
    }

    public async Task<Product> CreateProduct(string name, string description, decimal price, string category, int stockQuantity, string sku)
    {
        if (await _repository.ExistsBySku(sku))
            throw new DomainException($"A product with SKU '{sku}' already exists.");

        var product = Product.Create(name, description, price, category, stockQuantity, sku);
        return await _repository.Save(product);
    }

    public async Task<Product> GetProductById(Guid id)
    {
        var product = await _repository.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        return product;
    }

    public async Task<(IEnumerable<Product> Items, int TotalCount)> ListProducts(int page, int pageSize)
    {
        if (pageSize > 100) pageSize = 100;
        if (pageSize < 1) pageSize = 10;
        if (page < 0) page = 0;

        return await _repository.FindAll(page, pageSize);
    }

    public async Task<IEnumerable<Product>> SearchProducts(string query)
    {
        if (string.IsNullOrWhiteSpace(query))
            return Enumerable.Empty<Product>();

        return await _repository.Search(query);
    }

    public async Task<Product> UpdateProduct(Guid id, string name, string description, decimal price, string category, string sku)
    {
        var product = await _repository.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        if (await _repository.ExistsBySkuAndIdNot(sku, id))
            throw new DomainException($"A product with SKU '{sku}' already exists.");

        product.Update(name, description, price, category, sku);
        return await _repository.Update(product);
    }

    public async Task DeleteProduct(Guid id)
    {
        var product = await _repository.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        product.Deactivate();
        await _repository.Update(product);
    }

    public async Task<Product> DecreaseStock(Guid id, int quantity)
    {
        var product = await _repository.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        product.DecreaseStock(quantity);
        return await _repository.Update(product);
    }

    public async Task<Product> IncreaseStock(Guid id, int quantity)
    {
        var product = await _repository.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        product.IncreaseStock(quantity);
        return await _repository.Update(product);
    }
}

/// <summary>
/// Exception thrown when a product is not found.
/// </summary>
public class ProductNotFoundException : Exception
{
    public Guid ProductId { get; }

    public ProductNotFoundException(Guid id)
        : base($"Product with ID '{id}' was not found.")
    {
        ProductId = id;
    }
}
